Source/Request.swift

```diff
@@ -414,3 +323,283 @@ extension Request: CustomDebugStringConvertible {
+        case stream(InputStream, URLRequest)
+
+        func task(session: URLSession, adapter: RequestAdapter?, queue: DispatchQueue) -> URLSessionTask {
+            var task: URLSessionTask!
+
+            switch self {
+            case let .data(data, urlRequest):
+                let urlRequest = urlRequest.adapt(using: adapter)
+                queue.sync { task = session.uploadTask(with: urlRequest, from: data) }
+            case let .file(url, urlRequest):
+                let urlRequest = urlRequest.adapt(using: adapter)
+                queue.sync { task = session.uploadTask(with: urlRequest, fromFile: url) }
+            case let .stream(_, urlRequest):
+                let urlRequest = urlRequest.adapt(using: adapter)
+                queue.sync { task = session.uploadTask(withStreamedRequest: urlRequest) }
+            }
+
+            return task
+        }
+    }
+
+    // MARK: Properties
+
+    /// The progress of uploading the payload to the server for the upload request.
+    open var uploadProgress: Progress { return uploadDelegate.uploadProgress }
+
+    var uploadDelegate: UploadTaskDelegate { return delegate as! UploadTaskDelegate }
+
+    // MARK: Upload Progress
+
+    /// Sets a closure to be called periodically during the lifecycle of the `UploadRequest` as data is sent to 
+    /// the server.
+    ///
+    /// After the data is sent to the server, the `progress(queue:closure:)` APIs can be used to monitor the progress
+    /// of data being read from the server.
+    ///
+    /// - parameter queue:   The dispatch queue to execute the closure on.
+    /// - parameter closure: The code to be executed periodically as data is sent to the server.
+    ///
+    /// - returns: The request.
+    @discardableResult
+    open func uploadProgress(queue: DispatchQueue = DispatchQueue.main, closure: ProgressHandler) -> Self {
+        uploadDelegate.uploadProgressHandler = (closure, queue)
+        return self
+    }
+
+    /// Sets a closure to be called periodically during the lifecycle of the `UploadRequest` as data is sent to
+    /// the server.
+    ///
+    /// After the data is sent to the server, the `progress(queue:closure:)` APIs can be used to monitor the progress
+    /// of data being read from the server.
+    ///
+    /// - parameter queue:   The dispatch queue to execute the closure on.
+    /// - parameter closure: The code to be executed periodically as data is sent to the server.
+    ///
+    /// - returns: The request.
+    @discardableResult
+    open func uploadProgress(queue: DispatchQueue = DispatchQueue.main, closure: UploadProgressHandler) -> Self {
+        uploadDelegate.uploadProgressDebugHandler = (closure, queue)
+        return self
+    }
+}
+
+// MARK: -
+
+#if !os(watchOS)
+
+/// Specific type of `Request` that manages an underlying `URLSessionStreamTask`.
+open class StreamRequest: Request {
+    enum Streamable: TaskConvertible {
+        case stream(String, Int)
```

Comment:
Do we want labels on these cases? I know they're internal but it may be good even for us.